src/Sparrow/Json/BlittableJsonReaderObject.cs

```diff
@@ -483,5 +485,180 @@ public void CopyTo(byte* ptr)
+            currentSize -= offset;
+
+            if ((propsOffsetList > currentSize) || (propsOffsetList <= 0))
+                throw new InvalidDataException("Properties names offset not valid");
+
+            int propNamesOffsetSize;
+            var token = (BlittableJsonToken)(*(_mem + propsOffsetList));
+            propNamesOffsetSize = ProcessTokenOffsetFlags(token);
+            if (((token & (BlittableJsonToken)0xC0) != 0) || ((TypesMask & token) != 0x00))
+                throw new InvalidDataException("Properties names token not valid");
+
+            var numberOfProps = (currentSize - propsOffsetList) / propNamesOffsetSize;
+            currentSize = PropertiesNamesValidation(numberOfProps, propsOffsetList,
+                propNamesOffsetSize, propsOffsetList);
+
+            if ((rootMetadataOffset > currentSize) || (rootMetadataOffset < 0))
+                throw new InvalidDataException("Root metadata offset not valid");
+
+            var current = PropertiesValidation(rootToken, rootPropOffsetSize, rootPropIdSize,
+                rootMetadataOffset, numberOfProps);
+
+            if (current != currentSize)
+                throw new InvalidDataException("Root metadata not valid");
+        }
+
+        private int PropertiesNamesValidation(int numberOfProps, int propsOffsetList, int propsNamesOffsetSize,
+            int currentSize)
+        {
+            var offsetCounter = 0;
+            for (var i = numberOfProps; i > 0; i--)
+            {
+                int stringLength;
+                var nameOffset = 0;
+                nameOffset = ReadNumber((_mem + propsOffsetList + i * propsNamesOffsetSize),
+                    propsNamesOffsetSize);
+                if ((nameOffset > currentSize) || (nameOffset < 0))
+                    throw new InvalidDataException("Properties names offset not valid");
+                stringLength = StringValidation(propsOffsetList - nameOffset);
+                if (offsetCounter + stringLength + 1 != nameOffset)
+                    throw new InvalidDataException("Properties names offset not valid");
+                offsetCounter = nameOffset;
+                currentSize -= (stringLength + 1);
+            }
+            return currentSize;
+        }
+
+        private int StringValidation(int stringOffset)
+        {
+            byte offset;
+            int stringLength;
+            stringLength = ReadVariableSizeInt(stringOffset, out offset);
+            if (stringLength < 0)
+                throw new InvalidDataException("String not valid");
+            var str = stringOffset + offset;
+            var stringEnd = ReadVariableSizeInt(stringOffset + offset + stringLength, out offset);
+            if (stringEnd != 0)
+            {
+                for (var i = 0; i < stringEnd; i++)
+                {
+                    var escCharOffset = ReadNumber(_mem + str + stringLength + offset + i, 1);
```

Comment:
The variable name is misleading, `stringEnd` is actually the escape positions count.